Create CI

ilkerhalil opened the pull request:
- Create CI Step for Github Workflow
- Create CD Step for Github Workflow
- Fixed JsonParser "unfilledCount" error
- Fixed bufc compiled error.

Rjvs committed:
Minimal changes to readme

TYoungSL committed:
initial shot at 64-bit flatbuffers

TYoungSL committed:
regenerate monster sample

TYoungSL committed:
break returns on bfbs Deserialize path in idl_parser.cpp onto new lines for easier breakpoints

FIXME: commented out bfbs handling in sample_bfbs.cpp; it's broken

TYoungSL committed:
does the CI work?

TYoungSL committed:
regenerate reflection_generated.h

TYoungSL committed:
generate include_test files? is this a bad idea?

TYoungSL committed:
trying to use 64-bit indexing and sizing

TYoungSL committed:
have C# runtime use BigSpans

TYoungSL committed:
drop net 3.5 support

snyk-bot committed:
fix: upgrade io.grpc:grpc-core from 1.36.0 to 1.38.0

Snyk has created this PR to upgrade io.grpc:grpc-core from 1.36.0 to 1.38.0.

See this package in Maven Repository:
https://mvnrepository.com/artifact/io.grpc/grpc-core/

See this project in Snyk:
https://app.snyk.io/org/stirling-labs/project/1a437581-a9ef-4564-be5d-bd17d717ec39?utm_source=github&utm_medium=upgrade-pr

TYoungSL committed:
rework of C# FlatBuffers into BigBuffers

changed implementation of buffer allocation to work forward and perform less allocs

creates output in flatcc order now, uses placeholders to write back offsets

languages that will not be supported in this fork (yet)

use BigSpans

TYoungSL committed:
placeholder vector of inline structs

generated structs now know their size in bytes

IFlatbufferObject -> IBigBufferModel, IBigBufferStruct, IBigBufferTable

TYoungSL committed:
add single offset placeholder support

fix vtable location recording

make all c# structs emit as partial (may revert)

add more tests

start working on nested tables

TODO: existing vtable reference

TYoungSL committed:
fixed existing vtable usage

TYoungSL committed:
change placeholder offset value to long.MinValue, unify references to const

add tracker to Placeholder to help track debugging efforts

TYoungSL committed:
cumulative work toward flatcc compatible 64-bit offset size flat buffers

* add equality operators to tables and structs (IEquatable<T>)

* add readonly constraints to some buffer types to support fast reads and equality checks

* unify placeholder offset value used as BigBufferBuilder.PlaceholderOffset const

* restructure ByteBufferAllocator to expose Buffer as read-only

* add readonly decoration to methods that that do not manipulate the ByteBuffer

* update dependencies

made ByteBufferResidentModel, Struct and Table readonly

* made SchemaModel extensions work as readonly

* fixed generated emission of enum types as refs

* not generating reffers for unions right now

* add TypeEquals impl and tests

* expose internals of BigBuffers core to BigBuffers.Tests

* align and update type defs in idl.h to 64-bit types

TYoungSL committed:
add more unit tests

add Placeholder.ValidateAllFilled to run under debug only

add non-generic Offset type

flesh out placeholder tracking validation

TYoungSL committed:
automatically inferred generics cleanup

make UseExistingVTables path more explicitly chosen

TYoungSL committed:
rename BigBuffers project to BigBuffers.Runtime

complete implementation of tests over generated models

move __init and __assign out of models, add internal constructor to generated models

lots of fixes to offsets, placeholders, allocation, reading and writing to buffers

TYoungSL committed:
fix existing vtable check, added test

return field trimming check, needs test

TYoungSL committed:
some code cleanup

TYoungSL committed:
run against flatcc's monster

handle all sorts of weird scenarios

consolidate table and struct internal model representation

add support for generators reporting errors

use ValueTuples for inlined structs in factory (Create*) parameters

allow deprecating struct fields (but give a warning about them)

TYoungSL committed:
add MonsterDataTest

add TODOs to fix strings

fix verifying file identifier

TYoungSL committed:
fix prep for vectors and strings

fixed and improved some tests

made minimum buffer allocation 32 bytes

replaced some magic constants (8 -> sizeof(ulong))

improved TableE test to ensure alignment isn't covering up a missing null terminator

validate count of unfilled placeholders too

remove some leftover comments

TYoungSL committed:
string concatenation cleanup

set editor column size to 240 (may reduce to 120)

more clean up

make string creation default to shared

make the static __indirect method an extension method

TYoungSL committed:
create BigBuffers.JsonParsing

create tests

rename IBigBufferModel to IBigBufferEntity

add MetadataIndexAttribute, add UnionTypeAssociationAttribute

add Alignment to IBigBufferStruct

implement BigBufferBuilder.Begin from the remains of Finish, rework Finish

implement BigBufferBuilder.WriteOffset variations (avoiding collision with AddOffset)

add InternalsVisibleTo(BigBuffers.JsonParsing) assembly attribute

add OffsetExtensions.Resolve

add additional implicit operator to convert entities to offsets

add some Debug.Asserts

add and use InfoOf.Fody

add TableS for tests

fixed some x = target - offset vs x = offset - target stuff... cases for both usages fsr

update c# code generation, adds metadata, begin/finish, entity, placeholder marking, j -> _j

TYoungSL committed:
seems to be compatible with flatcc now

have Finish pass on the offset, we'll at least let the user shoot themselves

fix alignment issues everywhere

TYoungSL committed:
implement an alignment padding toggle

TYoungSL committed:
rename flatc to bufc

TYoungSL committed:
implement binary comparison for all bigbuf struct types

allow structs as keys

add GetByteSpan extension to all struct entity types

TYoungSL committed:
make key struct fields always required

make required struct fields not nullable

simplify key comparers

simplify struct comparers

add IsValid extension for entity types

TYoungSL committed:
drop rust support for now

TYoungSL committed:
add nuget.config

TYoungSL committed:
add differences documentation

ilkerhalil committed:
Fixed Project same error
- Build Bufc
- Create sln file for net project
- Fixed Monster Project
- Fixed Bufc source
- Fixed GenerateFixture Project

ilkerhalil committed:
Revert "Fixed Project same error"

This reverts commit f02acb5c4ae101179c4d417fd73999ae2c57fbf6.

ilkerhalil committed:
fixed bufc build

ilkerhalil committed:
add ci step for github action

ilkerhalil committed:
added BigBuffer net project for ci workflow

ilkerhalil committed:
added BigBuffer net project for ci workflow

ilkerhalil committed:
fixed find file step

ilkerhalil committed:
fixed restore nuget source

ilkerhalil committed:
fixed msbuid step

ilkerhalil committed:
delete version.proj

ilkerhalil committed:
change dotnet sln add strategy

ilkerhalil committed:
fixed windows build step

ilkerhalil committed:
fix copy articaft step

ilkerhalil committed:
add manuel csproj

ilkerhalil committed:
fixed create sln step

ilkerhalil committed:
try another build.

ilkerhalil committed:
add other step net code build

ilkerhalil committed:
fix linux and macos step

ilkerhalil committed:
another build try

ilkerhalil committed:
another try

ilkerhalil committed:
add build verbosity

ilkerhalil committed:
another try

ilkerhalil committed:
another try

ilkerhalil committed:
try retry

ilkerhalil committed:
try retry for windows worflow

ilkerhalil committed:
fix windows step

ilkerhalil committed:
fix windows step

ilkerhalil committed:
fix coverlet directory

ilkerhalil committed:
remove upload coverage step

ilkerhalil committed:
fixed windows artifact path

ilkerhalil committed:
fixed windows artifact path

ilkerhalil committed:
fix dll path

ilkerhalil committed:
list directory comm

ilkerhalil committed:
fix paths

ilkerhalil committed:
retry max attempts increase

ilkerhalil committed:
fix upload artifact versipn

ilkerhalil committed:
fix upload artifact versipn

ilkerhalil committed:
fixed nuget path

ilkerhalil committed:
add release wf for github

TYoungSL committed:
always create setters (they will be used by create steps)

declare them internal if mutable_buffer option is not set

TYoungSL committed:
remove some @ prefixes now that they're being generated in WrapInNameSpace

add case for IsFirstBuild detection in csproj where BuildInParallel is false (not sure if problem)

TYoungSL committed:
remove existing ci

TYoungSL committed:
update field name collision checks

ilkerhalil committed:
add sln file, change gitignore,

ilkerhalil committed:
fixed configure cmake step

ilkerhalil committed:
fixed cmake step

ilkerhalil committed:
fixed linux step

ilkerhalil committed:
fixed linux step

ilkerhalil committed:
fixed linux step

ilkerhalil committed:
fixed ci step

ilkerhalil committed:
fixed linux artifact path

ilkerhalil committed:
- Fixed JsonParser
- Fixed Windows cmake step inside directory.

ilkerhalil committed:
fixed build for windows bufc step

ilkerhalil committed:
- Fixed Windows bufc directory
- Fixed GeneratedFixture and Monster csproj bufc settings

TYoungSL committed:
add rpc pure interface codegen

add fbs csharp_value_task attribute for rpc_service and calls

update FluentAssertions

change MakeCamel to make first character lowercase if parameter first is false

add BigBuffers.Runtime.Xpc project

ilkerhalil committed:
fixed windows bufc artifact directory location

ilkerhalil committed:
complete cd wf, add Runtime.Xpc project,fixed artifact name

ilkerhalil committed:
fixed cd wf

ilkerhalil committed:
fixed artifact path for nuget push step

Rjvs commented on net/BigBuffers.Tests/BigBuffers.Tests.csproj:
```diff
@@ -6,6 +6,13 @@
 
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <Authors>Stirling Labs</Authors>
```
```suggestion
    <Authors>Stirling Labs Team & Contributors</Authors>
```

Rjvs reviewed (approved):
LGTM

ilkerhalil committed:
Changed -Authors-  BigBuffer test project

TYoungSL commented on net/BigBuffers.Tests.GeneratedFixture/BigBuffers.Tests.GeneratedFixture.csproj:
```diff
@@ -48,19 +48,19 @@
     <ItemGroup>
       <LocOfBufC Include="../../cmake-build-$(Configuration.ToLower())/bufc" />
-      <LocOfBufC Include="../../cmake-build-$(Configuration.ToLower())/bufc.exe" />
+      <LocOfBufC Include="../../cmake-build-$(Configuration.ToLower())/$(Configuration.ToLower())/bufc.exe" />
```
```suggestion
      <LocOfBufC Include="../../cmake-build-$(Configuration.ToLower())/bufc.exe" />
      <LocOfBufC Include="../../cmake-build-$(Configuration.ToLower())/$(Configuration.ToLower())/bufc.exe" />
```

TYoungSL commented on .github/workflows/cd.yml:
```diff
@@ -0,0 +1,118 @@
+          mkdir cmake-build-release && cd cmake-build-release && cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE=Release ..
+      - name: Build bufc
+        run: |
+          make bufc
```
This should probably be using `cmake --build` instead.

(Actual command line may vary, maybe `cmake --build .` or `cmake --build ..` or `cmake --build cmake-build-release` or something.)

TYoungSL commented on .github/workflows/cd.yml:
```diff
@@ -0,0 +1,118 @@
+          mkdir cmake-build-release && cd cmake-build-release && cmake -G "Visual Studio 16 2019" -A x64 -DCMAKE_BUILD_TYPE=Release ..
+      - name: Build bufc
+        run: |
+          msbuild.exe bufc.vcxproj /p:Configuration=Release /p:Platform=x64
```
This should probably be using `cmake --build .` instead.

TYoungSL commented on .github/workflows/cd.yml:
```diff
@@ -0,0 +1,118 @@
+        working-directory: cmake-build-release
+      - name: Update Nuget Source
+        run: |
+          dotnet nuget update source GitHub-StirlingLabs --source https://nuget.pkg.github.com/StirlingLabs/index.json --username ${{github.repository_owner}}  --password ${{github.token}} --store-password-in-clear-text
```
Should probably use `StirlingLabs/GithubNugetAuthAction` here instead.

TYoungSL commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,86 @@
+        working-directory: cmake-build-release
+      - name: Update Nuget Source
+        run: |
+          dotnet nuget update source GitHub-StirlingLabs --source https://nuget.pkg.github.com/StirlingLabs/index.json --username ${{github.repository_owner}}  --password ${{github.token}} --store-password-in-clear-text
```
Should probably use `StirlingLabs/GithubNugetAuthAction` here instead.

TYoungSL commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,86 @@
+          mkdir cmake-build-release && cd cmake-build-release && cmake -G "Visual Studio 16 2019" -A x64 -DCMAKE_BUILD_TYPE=Release ..
+      - name: Build bufc
+        run: |
+          msbuild.exe bufc.vcxproj /p:Configuration=Release /p:Platform=x64
```
This should probably be using `cmake --build .` instead.

TYoungSL commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,86 @@
+        working-directory: cmake-build-release
+      - name: Update Nuget Source
+        run: |
+          dotnet nuget update source GitHub-StirlingLabs --source https://nuget.pkg.github.com/StirlingLabs/index.json --username ${{github.repository_owner}}  --password ${{github.token}} --store-password-in-clear-text
```
Should probably use `StirlingLabs/GithubNugetAuthAction` here instead.

TYoungSL commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,86 @@
+          dotnet-version: 5.0.x
+      - name: Configure cmake
+        run: |
+          mkdir cmake-build-release && cd cmake-build-release && cmake -G "Visual Studio 16 2019" -A x64 -DCMAKE_BUILD_TYPE=Release ..
```
When building locally with MSys2 Makefiles (or Ninja under Msys2's CMake), there's no release subdirectory.

Visual Studio CMake targets are de facto multi-configuration environments, so they have Debug and Release output directories.

You'd have to add `-DCMAKE_RUNTIME_OUTPUT_DIRECTORY=${{github.workspace}}/cmake-build-release` to deal with that correctly.

It should be fine to leave it as is, but note the change in the project to add a search location instead of changing the search location.

TYoungSL commented on net/BigBuffers.Tests.Monster/BigBuffers.Tests.Monster.csproj:
```diff
@@ -54,33 +54,33 @@
     <ItemGroup>
       <LocOfBufC Include="../../cmake-build-$(Configuration.ToLower())/bufc" />
-      <LocOfBufC Include="../../cmake-build-$(Configuration.ToLower())/bufc.exe" />
+      <LocOfBufC Include="../../cmake-build-$(Configuration.ToLower())/$(Configuration.ToLower())/bufc.exe" />
```
```suggestion
      <LocOfBufC Include="../../cmake-build-$(Configuration.ToLower())/bufc.exe" />
      <LocOfBufC Include="../../cmake-build-$(Configuration.ToLower())/$(Configuration.ToLower())/bufc.exe" />
```

This should be adding to the search locations instead of changing the existing ones.

TYoungSL commented on src/idl_gen_cpp.cpp:
```diff
@@ -3166,7 +3166,7 @@ class CppGenerator : public BaseGenerator {
       const std::function<void(int bits, std::string *code_ptr, int *id)> &f) {
     if (field.padding) {
-      for (int i = 0; i < sizeof(uoffset_t); i++) {
+      for (uoffset_t i = 0; i < sizeof(uoffset_t); i++) {
```
```suggestion
      for (size_t i = 0; i < sizeof(uoffset_t); i++) {
```

TYoungSL replied:
This is in main now. 👍 
Can rebase and discard conflicting change to uoffset_t.

TYoungSL commented on src/idl_gen_csharp.cpp:
```diff
@@ -2561,7 +2561,7 @@ class CSharpGenerator : public BaseGenerator {
               "  public string SerializeToJson() {\n"
               "    return Newtonsoft.Json.JsonConvert.SerializeObject(this, "
-              "Newtonsoft.Json.Formatting.Indented);\n";
+              "Newtonsoft.Json.Formatting.Indented);\n"
```
Good catch.

TYoungSL replied:
This is in main now.

TYoungSL committed:
fix some gotchas VS doesn't complain about, thanks ilker!

TYoungSL commented on .github/workflows/cd.yml:
```diff
@@ -0,0 +1,118 @@
+        working-directory: cmake-build-release
+      - name: Update Nuget Source
+        run: |
+          dotnet nuget update source GitHub-StirlingLabs --source https://nuget.pkg.github.com/StirlingLabs/index.json --username ${{github.repository_owner}}  --password ${{github.token}} --store-password-in-clear-text
```
Should probably use `StirlingLabs/GithubNugetAuthAction` here instead.

Rjvs replied:
Good catch

TYoungSL commented on .github/workflows/cd.yml:
```diff
@@ -0,0 +1,118 @@
+          dotnet-version: 5.0.x
+      - name: Configure cmake
+        run: |
+          mkdir cmake-build-release && cd cmake-build-release && cmake -G "Visual Studio 16 2019" -A x64 -DCMAKE_BUILD_TYPE=Release ..
```
When building locally with MSys2 Makefiles (or Ninja under Msys2's CMake), there's no release subdirectory.

Visual Studio CMake targets are de facto multi-configuration environments, so they have Debug and Release output directories.

You'd have to add `-DCMAKE_RUNTIME_OUTPUT_DIRECTORY=${{github.workspace}}/cmake-build-release` to deal with that correctly.

It should be fine to leave it as is, but note the change in the project to add a search location instead of changing the search location.

TYoungSL commented on net/BigBuffers.JsonParsing/JsonParser.cs:
```diff
@@ -541,8 +541,9 @@ public T Parse(JsonElement element)
           while (DeferredActions.TryTake(out var filler))
             filler();
-
+#if DEBUG
           Placeholder.ValidateUnfilledCount(Builder, unfilledCount);
+#endif
```
```suggestion
          Placeholder.ValidateUnfilledCount(Builder, unfilledCount);
```

TYoungSL replied:
ValidateUnfilledCount is already marked `[Conditional("DEBUG")]` - doesn't need #if DEBUG around it.

TYoungSL commented on net/BigBuffers.Tests/BigBuffers.Tests.csproj:
```diff
@@ -6,6 +6,13 @@
+    <Authors>Stirling Labs Team &amp; Contributors</Authors>
+    <CollectCoverage>true</CollectCoverage>
+    <CoverletOutputFormat>cobertura</CoverletOutputFormat>
+    <CoverletOutput>../../coverage/</CoverletOutput>
```
```suggestion
    <CoverletOutput>../coverage/</CoverletOutput>
```

This should probably just be for the solution directory; there will be other coverage information from other projects.

TYoungSL commented on net/BigBuffers.Tests/BigBuffers.Tests.csproj:
```diff
@@ -15,11 +22,14 @@
-    <PackageReference Include="StirlingLabs.BigSpans.NUnit" Version="21.8.1" />
+    <PackageReference Include="coverlet.msbuild" Version="3.1.0" PrivateAssets="all" />
+    <PackageReference Include="ReportGenerator" Version="4.8.12" PrivateAssets="all" />
+    <PackageReference Include="StirlingLabs.BigSpans.NUnit" Version="21.7.0" />
```
```suggestion
    <PackageReference Include="StirlingLabs.BigSpans" Version="21.8.1" />
    <PackageReference Include="StirlingLabs.BigSpans.NUnit" Version="21.8.1" />
```
Old version. More correct to include a dependency directly on `StirlingLabs.BigSpans` too.

TYoungSL commented on net/BigBuffers.Tests/BigBuffers.Tests.csproj:
```diff
@@ -28,5 +38,8 @@
-
+  <Target Name="GenerateHtmlCoverageReport" AfterTargets="GenerateCoverageResultAfterTest">
+    <!-- generate markdown coverage report and badges -->
+    <ReportGenerator ReportFiles="@(CoverletReport)" ReportTypes="MarkdownSummary;Badges" SourceDirectories="../BigBuffers.Runtime" TargetDirectory="../../coverage/" />
```
```suggestion
    <ReportGenerator ReportFiles="@(CoverletReport)" ReportTypes="MarkdownSummary;Badges" SourceDirectories="../BigBuffers.Runtime" TargetDirectory="../coverage/" />
```
Just up to the solution directory instead of to the repository root.

TYoungSL commented on .gitignore:
```diff
@@ -17,7 +17,7 @@ build
 *.vcxproj.filters
 *.vcxproj.user
-*.sln
+#*.sln
```
```suggestion
```

TYoungSL commented on .gitignore:
```diff
@@ -17,7 +17,7 @@ build
 *.vcxproj
 *.vcxproj.filters
```
```suggestion
```

TYoungSL commented on .gitignore:
```diff
@@ -138,4 +138,7 @@ yarn-error.log
 **/FlatBuffers.Test.Swift.xcodeproj
-/cmake-build-*
\ No newline at end of file
+/cmake-build-*
```
```suggestion
cmake-build-*/
```

TYoungSL commented on .gitignore:
```diff
@@ -138,4 +138,7 @@ yarn-error.log
-/cmake-build-*
\ No newline at end of file
+/cmake-build-*
+bufc
+bufc.exe
```
```suggestion
```

TYoungSL commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "Version.Net"]
+	path = Version.Net
+	url = git@github.com:StirlingLabs/Version.Net.git
```
Please move the submodule from the repo root to the solution directory, `/Version.Net` -> `/net/Version.Net`

TYoungSL reviewed (changes requested):
Version.Net was kind of out of scope for this project at the moment, but might as well. 👍 

Please move the Version.Net submodule to the `net/` subdirectory.

This repository contains multiple non-.NET language projects.

Per docs for Version.Net,
> the solution level is assumed to be the directory of the solution (.sln) file, typically at the repository root

In this case the solution directory is `net/` instead of the repository root at `/`.

TYoungSL committed:
woops

ilkerhalil committed:
Fixed bug that occurred during rebase for BigBuffers.Tests.GeneratedFixture.csproj file

ilkerhalil committed:
Fixed cd workflow

ilkerhalil committed:
Changed Nuget Update step GithubNugetAuthAction
Changed Build bufc for windows step
Added Configure cmake step output directory for windows
Removed JsonParser.cs in 'endif' directive
Changed coverage output directory
Updated StirlingLabs.BigSpans.NUnit package version
Fixed .gitignore

ilkerhalil committed:
Changed "Update Nuget Source" step GithubNugetAuthAction for cd workflow
Changed "Build bufc" step cmake for ci workflow

ilkerhalil committed:
Changed GithubNugetAuthAction token

ilkerhalil committed:
changed cmake target to bufc
